man/groupmems.8
.\"     Title: groupmems
.\"    Author: 
.\" Generator: DocBook XSL Stylesheets v1.70.1 <http://docbook.sf.net/>
.\"      Date: 07/30/2006
.\"    Manual: System Management Commands
.\"    Source: System Management Commands
.\"
.TH "GROUPMEMS" "8" "07/30/2006" "System Management Commands" "System Management Commands"
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.SH "NAME"
groupmems \- administer members of a user's primary group
.SH "SYNOPSIS"
.HP 10
\fBgroupmems\fR \-a\ \fIuser_name\fR | \-d\ \fIuser_name\fR | \-l | \-p | [\-g\ \fIgroup_name\fR] 
.SH "DESCRIPTION"
.PP
The
\fBgroupmems\fR
utility allows a user to administer his/her own group membership list without the requirement of super user privileges. The
\fBgroupmems\fR
utility is for systems that configure its users to be in their own name sake primary group (i.e., guest / guest).
.PP
Only the super user, as administrator, can use
\fBgroupmems\fR
to alter the memberships of other groups.
.SH "OPTIONS"
.PP
The options which apply to the
\fBgroupmems\fR
command are:
.TP 3n
\fB\-a\fR \fIuser_name\fR
Add a new user to the group membership list.
.TP 3n
\fB\-d\fR \fIuser_name\fR
Delete a user from the group membership list.
.TP 3n
\fB\-p\fR
Purge all users from the group membership list.
.TP 3n
\fB\-g\fR \fIgroup_name\fR
The super user can specify which group membership list to modify.
.TP 3n
\fB\-l\fR
List the group membership list.
.SH "SETUP"
.PP
The
\fBgroupmems\fR
executable should be in mode
2770
as user
\fIroot\fR
and in group
\fIgroups\fR. The system administrator can add users to group
\fIgroups\fR
to allow or disallow them using the
\fBgroupmems\fR
utility to manage their own group membership list.
.sp
.RS 3n
.nf
	$ groupadd \-r groups
	$ chmod 2770 groupmems
	$ chown root.groups groupmems
	$ groupmems \-g groups \-a gk4
    
.fi
.RE
.SH "FILES"
.TP 3n
\fI/etc/group\fR
Group account information.
.TP 3n
\fI/etc/gshadow\fR
secure group account information
.SH "SEE ALSO"
.PP

\fBchfn\fR(1),
\fBchsh\fR(1),
\fBpasswd\fR(1),
\fBgroupadd\fR(8),
\fBgroupdel\fR(8),
\fBuseradd\fR(8),
\fBuserdel\fR(8),
\fBusermod\fR(8).


NEWS
$Id: NEWS,v 1.492 2007/02/01 20:49:25 kloczek Exp $

shadow-4.0.18.1 -> shadow-4.0.18.2					13-10-2007

*** general:
- usermod: fixed handle -a option (by Benno Schulenberg
  <bensberg@justemail.net>),
- useradd: improved auditing support
  (https://bugzilla.redhat.com/bugzilla/show_bug.cgi?id=211659),
- groupadd, groupdel, groupmod, useradd, userdel, usermod: flush nscd cashes
  after close /etc/{group,passwd} files,
- su: If compiled without PAM support, enforce the limits from /etc/limits
  when one of the -, -l, or --login options is set, even if called by root.
*** documentation:
- updated translations: ja, nl, tl.
- groupadd.8, groupmod.8, login.1, useradd.8, userdel.8, usermod.8: grammar
  mistakes and other correctstions (by Schulenberg <bensberg@justemail.net>),

shadow-4.0.18 -> shadow-4.0.18.1					03-08-2006

*** general:
- groupmems: fixed compilation when PAM is disabled
  (by Johannes Winkelmann <jw@smts.ch>),
- fixed missing man pages in dist tar ball necessary on build when
  PAM is disabled.

shadow-4.0.17 -> shadow-4.0.18						01-08-2006

*** general:
- su: fixed set enviroment too early when using PAM, so move it to !USE_PAM
  (patch submitted by Mike Frysinger <vapier@gentoo.org>),
- groupadd, groupmod, useradd, usermod: fixed UID/GID overflow (fixed
  http://bugzilla.redhat.com/bugzilla/show_bug.cgi?id=198920)
- passwd, useradd, usermod: fixed inactive/mindays/warndays/maxdays overflow
  (simillar to RH#198920),
- groupmems: rewrited for use PAM and getopt_long() and now it is enabled
  for build and install (patch by George Kraft <gk4@swbell.net>),
- S/Key: removed assign getpass() to libshadow_getpass() on autoconf level
  (patch by Ulrich Mueller <ulm@kph.uni-mainz.de>; http://bugs.gentoo.org/139966),
- usermod: back to previous -a option semantics and clarify -a behavior
  on documentation level (by Greg Schafer <gschafer@zip.com.au>),
- chsh, groupmod: rewrited for use getopt_long().
- updated translations: ca, cs, da, eu, fr, gl, hu, ko, pl, pt, ru, sv, tr, uk, vi.
*** documentation:
- fr and ru man pages are up to date,
- partially translated sv man pages set added
  (by Daniel Nylander <info@danielnylander.se>),
- pl chage(1), chsh(1), groupmod(8): translation finished.

shadow-4.0.16 -> shadow-4.0.17						10-07-2006

*** general:
- userdel, usermod: fixed segfault on remove home directory when it can't
  be removed; for example when it is /dev/null (fixed http://bugs.gentoo.org/139148),
- improved SELinux detection on autoconf level (based on patch by
  Dan Yefimov <dan@D00M.lightwave.net.ru>),
- removed using private implementation getpass() libc function
  (now getpass() is used also when S/KEY support is enabled),
- move nologin do $(sbindir),
- useradd: fixed mail spool file creation (bug cached by Frans Pop
  <elendil@planet.nl>;
  fixed http://bugs.debian.org/374705),
- updated translations: cs, da, de, ko, nb, nl, pt, ro, ru, sk, sv, vi,
- new translations: dz, km, ne.
*** documentation:
- ru man pages up to date,
- lastlog(8): updated pl translation,
- faillog(5): added missing information about fail_locktime element of
  faillog struct (by Thorsten Kukuk <kukuk@suse.de>),
- updated translations: eu, fr, pl.
- reverted using docbook.sourceforge.net in XSL url.

shadow-4.0.15 -> shadow-4.0.16						05-06-2006

*** general:
- userdel: better fix for old CERT VU#312962 (which was fixed in shadow 4.0.8):
  fixed forgoten checking of the return value from fchown() before
  proceeding with the fchmod() (based on Owl patch prepared by
  Rafal Wojtczuk <nergal@owl.openwall.com>),
- userdel: use login.defs::MAIL_DIR instead hardcoded /var/mail in created
  mailbox path (based on Owl fixes submited
  by Solar Designer <solar@openwall.com>),
- by default do not use libshadow_getpass() as getpass() replacemement.
  Use libshadow_getpass() only when S/KEY support is enabled.
  Current glibc getpass() handles correctly longer than 8 characters
  passwords and libshadow_getpass() is used only because libc getpass()
  do not handles password prompting with echo enabled,
- move login.defs::MD5_CRYPT_ENAB to non-PAM part,
- userdel: rewrited for use getopt_log(),
- install default/template configuration files:
-- if shadow is configured with use PAM install /etc/pam.d/* files,
-- if shadow do not uses PAM install /etc/{limits,login.acces} files,
-- install /etc/login.defs and /etc/default/useradd files,
- fixed handle relative symlinks too in lib/commonio.c
  (merge patch from Fedora),
- properly notify nscd to flush its cache
  (https://bugzilla.redhat.com/bugzilla/186803),
- useradd, usermod: fixes for verify return values mkdir() and chown()
  on copy files (merge 482_libmisc_copydir_check_return_values Debian
  patch),
- login, su (non-PAM variant): export MAIL only when MAIL_CHECK_ENAB
  is enabled (Mike Frysinger <vapier@gentoo.org>),
- pgck, grpck: warn when the members of a group differ in /etc/groups
  and /etc/gshadow (fixed http://bugs.debian.org/75181),
- su: fixed exit with a status 0 when the invoked command is terminated
  by a signal which was not catched
  (fixed by Eero Häkkinen <eero17@bigfoot.com>),
- login: cancel login timeout after authentication so that patient people
  timing out on network directory services can log in with local
  accounts (http://bugs.debian.org/107148),
- chgpasswd: fixes for build correctly with --disable-shadowgrp
  (patch by Johannes Winkelmann <jw@tks6.net>).
- updated translations: cs, da, es, eu, fi, fr, gl, hu, id, pt, ru, sk, sv, vi.
- new translations: hu.
*** documentation:
- new cs man pages: groupmems(8), groupmod(8), grpck(8), gshadow(5)
  (by Miroslav Kure <kurem@upcase.inf.upol.cz>),
- regenerate roff man pages using docbook-style-xsl-1.70.1,
- bunch of cleanups in chfn(1), faillog(8), gpasswd(1), groupadd(8),
  groupmems(8), limits(5), login(1), login.defs(5), newgrp(1), passwd(1),
  passwd(5) and su(1) (by Yuri Kozlov <kozlov.y@gmail.com>),
- update pl vipw(8) man page,
- added chgpasswd(8) ru man page,

src/groupmems.c
/*
 * Copyright 2000, International Business Machines
 * George Kraft IV, gk4@us.ibm.com, 03/23/2000
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. Neither the name of Julianne F. Haugh nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY JULIE HAUGH AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL JULIE HAUGH OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
*/

#include <config.h>

#include <fcntl.h>
#include <getopt.h>
#include <grp.h>
#include <stdio.h>
#include <sys/types.h>
#ifdef USE_PAM
#include "pam_defs.h"
#endif				/* USE_PAM */
#include <pwd.h>
#include "defines.h"
#include "groupio.h"

/* Exit Status Values */

#define EXIT_SUCCESS		0	/* success */
#define EXIT_USAGE		1	/* invalid command syntax */
#define EXIT_GROUP_FILE		2	/* group file access problems */
#define EXIT_NOT_ROOT		3	/* not superuser  */
#define EXIT_NOT_EROOT		4	/* not effective superuser  */
#define EXIT_NOT_PRIMARY	5	/* not primary owner of group  */
#define EXIT_NOT_MEMBER		6	/* member of group does not exist */
#define EXIT_MEMBER_EXISTS	7	/* member of group already exists */

#define TRUE 1
#define FALSE 0

/*
 * Global variables
 */
extern int optind;
extern char *optarg;
static char *adduser = NULL;
static char *deluser = NULL;
static char *thisgroup = NULL;
static int purge = FALSE;
static int list = FALSE;
static int exclusive = 0;

static int isroot (void)
{
	return getuid ()? FALSE : TRUE;
}

static int isgroup (void)
{
	gid_t g = getgid ();
	struct group *grp = getgrgid (g);

	return TRUE;
}

static char *whoami (void)
{
	struct group *grp = getgrgid (getgid ());
	struct passwd *usr = getpwuid (getuid ());

	if (0 == strcmp (usr->pw_name, grp->gr_name)) {
		return (char *) strdup (usr->pw_name);
	} else {
		return NULL;
	}
}

static void addtogroup (char *user, char **members)
{
	int i;

	for (i = 0; NULL != members[i]; i++) {
		if (0 == strcmp (user, members[i])) {
			fprintf (stderr, _("Member already exists\n"));
			exit (EXIT_MEMBER_EXISTS);
		}
	}

	members = (char **) realloc (members, sizeof (char *) * i);
	members[i] = user;
	members[i + 1] = NULL;
}

static void rmfromgroup (char *user, char **members)
{
	int i;
	int found = FALSE;

	i = 0;
	while (!found && NULL != members[i]) {
		if (0 == strcmp (user, members[i])) {
			found = TRUE;
		} else {
			i++;
		}
	}

	while (found && NULL != members[i]) {
		members[i] = members[++i];
	}

	if (!found) {
		fprintf (stderr, _("Member to remove could not be found\n"));
		exit (EXIT_NOT_MEMBER);
	}
}

static void nomembers (char **members)
{
	int i;

	for (i = 0; NULL != members[i]; i++) {
		members[i] = NULL;
	}
}

static void members (char **members)
{
	int i;

	for (i = 0; NULL != members[i]; i++) {
		printf ("%s ", members[i]);

		if (NULL == members[i + 1]) {
			printf ("\n");
		} else {
			printf (" ");
		}
	}
}

static void usage (void)
{
	fprintf (stderr,
		 _
		 ("Usage: groupmems -a username | -d username | -D | -l [-g groupname]\n"));
	exit (EXIT_USAGE);
}

int main (int argc, char **argv) 
{
	int arg;
	char *name;
	struct group *grp;

#ifdef USE_PAM
	pam_handle_t *pamh = NULL;
	struct passwd *pampw;
	int retval;
#endif

	setlocale (LC_ALL, "");
	bindtextdomain (PACKAGE, LOCALEDIR);
	textdomain (PACKAGE);

	int option_index = 0;
	static struct option long_options[] = {
		{"add", required_argument, NULL, 'a'},
		{"delete", required_argument, NULL, 'd'},
		{"group", required_argument, NULL, 'g'},
		{"list", no_argument, NULL, 'l'},
		{"purge", no_argument, NULL, 'p'},
		{NULL, 0, NULL, '\0'}
	};

	while ((arg =
		getopt_long (argc, argv, "a:d:g:lp", long_options,
			     &option_index)) != EOF) {
		switch (arg) {
		case 'a':
			adduser = strdup (optarg);
			++exclusive;
			break;
		case 'd':
			deluser = strdup (optarg);
			++exclusive;
			break;
		case 'p':
			purge = TRUE;
			++exclusive;
			break;
		case 'g':
			thisgroup = strdup (optarg);
			break;
		case 'l':
			list = TRUE;
			++exclusive;
			break;
		default:
			usage ();
		}
	}

	if (exclusive > 1 || optind < argc) {
		usage ();
	}

	if (!isroot () && NULL != thisgroup) {
		fprintf (stderr,
			 _("Only root can add members to different groups\n"));
		exit (EXIT_NOT_ROOT);
	} else if (isroot () && NULL != thisgroup) {
		name = thisgroup;
	} else if (!isgroup ()) {
		fprintf (stderr, _("Group access is required\n"));
		exit (EXIT_NOT_EROOT);
	} else if (NULL == (name = whoami ())) {
		fprintf (stderr, _("Not primary owner of current group\n"));
		exit (EXIT_NOT_PRIMARY);
	}
#ifdef USE_PAM
	retval = PAM_SUCCESS;

	pampw = getpwuid (getuid ());
	if (pampw == NULL) {
		retval = PAM_USER_UNKNOWN;
	}

	if (retval == PAM_SUCCESS) {
		retval = pam_start ("groupmod", pampw->pw_name, &conv, &pamh);
	}

	if (retval == PAM_SUCCESS) {
		retval = pam_authenticate (pamh, 0);
		if (retval != PAM_SUCCESS) {
			pam_end (pamh, retval);
		}
	}

	if (retval == PAM_SUCCESS) {
		retval = pam_acct_mgmt (pamh, 0);
		if (retval != PAM_SUCCESS) {
			pam_end (pamh, retval);
		}
	}

	if (retval != PAM_SUCCESS) {
		fprintf (stderr, _("PAM authentication failed for\n"));
		exit (1);
	}
#endif

	if (!gr_lock ()) {
		fprintf (stderr, _("Unable to lock group file\n"));
		exit (EXIT_GROUP_FILE);
	}

	if (!gr_open (O_RDWR)) {
		fprintf (stderr, _("Unable to open group file\n"));
		exit (EXIT_GROUP_FILE);
	}

	grp = (struct group *) gr_locate (name);

	if (NULL != adduser) {
		addtogroup (adduser, grp->gr_mem);
		gr_update (grp);
	} else if (NULL != deluser) {
		rmfromgroup (deluser, grp->gr_mem);
		gr_update (grp);
	} else if (purge) {
		nomembers (grp->gr_mem);
		gr_update (grp);
	} else if (list) {
		members (grp->gr_mem);
	}

	if (!gr_close ()) {
		fprintf (stderr, _("Cannot close group file\n"));
		exit (EXIT_GROUP_FILE);
	}

	gr_unlock ();

#ifdef USE_PAM
	if (retval == PAM_SUCCESS) {
		pam_end (pamh, PAM_SUCCESS);
	}
#endif				/* USE_PAM */
	exit (EXIT_SUCCESS);
}
